shared/actions/favorite.js

```diff
@@ -118,46 +120,85 @@ const folderToState = (folders: Array<FolderWithMeta>, username: string = ''): S
 // If the notify data has changed, show a popup
 let previousNotifyState = null
 
-export function favoriteList (): (dispatch: Dispatch) => void {
+const injectMeta = type => f => { f.meta = type }
+
+export function favoriteList (): (dispatch: Dispatch, getState: () => Object) => void {
   return (dispatch, getState) => {
-    favoriteGetFavoritesRpc({
-      callback: (error, favorites) => {
+    apiserverGetRpc({
+      param: {
+        endpoint: 'kbfs/favorite/list',
+        args: [{key: 'problems', value: '1'}],
+      },
+      callback: (error, result) => {
         if (error) {
-          console.warn('Err in favorite.getFavorites', error)
+          console.warn('Err in getFavorites', error)
           return
         }
 
-        let folders = []
+        let json
+        try {
+          json = JSON.parse(result.body)
+        } catch (err) {
+          console.warn('Invalid json from getFavorites: ', err)
+          return
+        }
 
-        if (favorites) {
-          if (favorites.favoriteFolders) {
-            favorites.favoriteFolders.forEach(f => {
-              folders.push({folder: f, meta: null})
-            })
-          }
+        const myUsername = getState().config && getState().config.username
+        const myKID = _.findKey(json.users, name => name === myUsername)
 
-          if (favorites.newFolders) {
-            favorites.newFolders.forEach(f => {
-              folders.push({folder: f, meta: 'new'})
-            })
-          }
+        // inject our meta tag
+        json.favorites && json.favorites.forEach(injectMeta(null))
+        json.ignored && json.ignored.forEach(injectMeta('ignored'))
+        json.new && json.new.forEach(injectMeta('new'))
 
-          if (favorites.ignoredFolders) {
-            favorites.ignoredFolders.forEach(f => {
-              folders.push({folder: f, meta: 'ignored'})
-            })
-          }
-        }
+        // figure out who can solve the rekey
+        const folderSets = [json.favorites, json.ignored, json.new]
```

Comment:
🐑 Maybe we can move this chunk of code to a separate function, to clean this up a bit?